// @ghmc/mobile/src/navigation/AppNavigator.tsx
import React from 'react'
import { View, Text, TextInput, TouchableOpacity, StyleSheet } from 'react-native'
import { NavigationContainer } from '@react-navigation/native'
import { createNativeStackNavigator } from '@react-navigation/native-stack'
import { createBottomTabNavigator } from '@react-navigation/bottom-tabs'
import { useAuthStore } from '../store/authStore'

const Stack = createNativeStackNavigator()
const Tab   = createBottomTabNavigator()

// ─── Simple placeholder ───────────────────────────────────────
function Placeholder({ label }: { label: string }) {
  return (
    <View style={s.center}>
      <Text style={s.placeholderText}>{label}</Text>
    </View>
  )
}

// ─── Tenant slug screen ───────────────────────────────────────
function TenantSlugScreen({ navigation }: any) {
  const [slug, setSlug] = React.useState('')
  const setTenantSlug   = useAuthStore((st: any) => st.setTenantSlug)

  return (
    <View style={s.authWrap}>
      <View style={s.logo} />
      <Text style={s.appTitle}>GHMC Campaign</Text>
      <Text style={s.appSub}>Election management platform</Text>
      <Text style={s.fieldLabel}>Campaign slug</Text>
      <TextInput
        style={s.input}
        value={slug}
        onChangeText={setSlug}
        placeholder="e.g. bjp-ward42"
        placeholderTextColor="#aaa"
        autoCapitalize="none"
        autoCorrect={false}
      />
      <Text style={s.hint}>Provided by your campaign coordinator</Text>
      <TouchableOpacity
        style={s.btn}
        onPress={() => {
          if (!slug.trim()) return
          setTenantSlug(slug.trim().toLowerCase())
          navigation.navigate('Phone')
        }}>
        <Text style={s.btnText}>Continue</Text>
      </TouchableOpacity>
    </View>
  )
}

// ─── Phone screen ─────────────────────────────────────────────
function PhoneScreen({ navigation }: any) {
  const [phone, setPhone]   = React.useState('')
  const requestOtp          = useAuthStore((st: any) => st.requestOtp)
  const isLoading           = useAuthStore((st: any) => st.isLoading)
  const error               = useAuthStore((st: any) => st.error)
  const tenantSlug          = useAuthStore((st: any) => st.tenantSlug)

  const handleSend = async () => {
    const cleaned = phone.replace(/\D/g, '')
    if (cleaned.length < 10) return
    try {
      await requestOtp(cleaned, tenantSlug)
      navigation.navigate('Otp', { phone: cleaned })
    } catch {}
  }

  return (
    <View style={s.authWrap}>
      <TouchableOpacity onPress={() => navigation.goBack()} style={s.backBtn}>
        <Text style={s.backText}>← Back</Text>
      </TouchableOpacity>
      <Text style={s.authTitle}>Enter your phone</Text>
      <Text style={s.authSub}>{tenantSlug} campaign</Text>
      {error ? <Text style={s.errorText}>{error}</Text> : null}
      <Text style={s.fieldLabel}>Mobile number</Text>
      <TextInput
        style={s.input}
        value={phone}
        onChangeText={setPhone}
        placeholder="9876543210"
        placeholderTextColor="#aaa"
        keyboardType="phone-pad"
        maxLength={10}
      />
      <TouchableOpacity style={[s.btn, isLoading && s.btnDisabled]} onPress={handleSend} disabled={isLoading}>
        <Text style={s.btnText}>{isLoading ? 'Sending...' : 'Send OTP'}</Text>
      </TouchableOpacity>
    </View>
  )
}

// ─── OTP screen ───────────────────────────────────────────────
function OtpScreen({ route, navigation }: any) {
  const { phone }   = route.params
  const [otp, setOtp] = React.useState('')
  const verifyOtp   = useAuthStore((st: any) => st.verifyOtp)
  const isLoading   = useAuthStore((st: any) => st.isLoading)
  const error       = useAuthStore((st: any) => st.error)

  const handleVerify = async () => {
    if (otp.length !== 6) return
    try { await verifyOtp(phone, otp) } catch {}
  }

  return (
    <View style={s.authWrap}>
      <TouchableOpacity onPress={() => navigation.goBack()} style={s.backBtn}>
        <Text style={s.backText}>← Back</Text>
      </TouchableOpacity>
      <Text style={s.authTitle}>Verify OTP</Text>
      <Text style={s.authSub}>Sent to +91 {phone}</Text>
      {error ? <Text style={s.errorText}>{error}</Text> : null}
      <Text style={s.fieldLabel}>6-digit code</Text>
      <TextInput
        style={[s.input, s.otpInput]}
        value={otp}
        onChangeText={setOtp}
        placeholder="• • • • • •"
        placeholderTextColor="#aaa"
        keyboardType="number-pad"
        maxLength={6}
      />
      <TouchableOpacity style={[s.btn, isLoading && s.btnDisabled]} onPress={handleVerify} disabled={isLoading}>
        <Text style={s.btnText}>{isLoading ? 'Verifying...' : 'Verify'}</Text>
      </TouchableOpacity>
    </View>
  )
}

// ─── Main tabs (post-login) ───────────────────────────────────
function MainTabs() {
  return (
    <Tab.Navigator screenOptions={{
      headerShown: false,
      tabBarActiveTintColor: '#1D9E75',
      tabBarInactiveTintColor: '#888',
      tabBarStyle: { paddingBottom: 6, height: 58 },
    }}>
      <Tab.Screen name="Home"    options={{ tabBarLabel: 'Home' }}>
        {() => <Placeholder label="Dashboard — loading..." />}
      </Tab.Screen>
      <Tab.Screen name="Voters"  options={{ tabBarLabel: 'Voters' }}>
        {() => <Placeholder label="Voters — loading..." />}
      </Tab.Screen>
      <Tab.Screen name="Tasks"   options={{ tabBarLabel: 'Tasks' }}>
        {() => <Placeholder label="Tasks — loading..." />}
      </Tab.Screen>
      <Tab.Screen name="Profile" options={{ tabBarLabel: 'Profile' }}>
        {() => <Placeholder label="Profile — loading..." />}
      </Tab.Screen>
    </Tab.Navigator>
  )
}

// ─── Root navigator ───────────────────────────────────────────
export function AppNavigator() {
  const user = useAuthStore((st: any) => st.user)
  return (
    <NavigationContainer>
      <Stack.Navigator screenOptions={{ headerShown: false }}>
        {user ? (
          <Stack.Screen name="Main" component={MainTabs} />
        ) : (
          <>
            <Stack.Screen name="TenantSlug" component={TenantSlugScreen} />
            <Stack.Screen name="Phone"      component={PhoneScreen} />
            <Stack.Screen name="Otp"        component={OtpScreen} />
          </>
        )}
      </Stack.Navigator>
    </NavigationContainer>
  )
}

const s = StyleSheet.create({
  center:          { flex: 1, justifyContent: 'center', alignItems: 'center', backgroundColor: '#fff' },
  placeholderText: { fontSize: 16, color: '#888' },
  authWrap:        { flex: 1, padding: 24, paddingTop: 80, backgroundColor: '#fff' },
  logo:            { width: 56, height: 56, borderRadius: 16, backgroundColor: '#1D9E75',
                     alignSelf: 'center', marginBottom: 16 },
  appTitle:        { fontSize: 22, fontWeight: '600', textAlign: 'center', color: '#1a1a18' },
  appSub:          { fontSize: 13, color: '#888', textAlign: 'center', marginBottom: 40 },
  authTitle:       { fontSize: 20, fontWeight: '600', color: '#1a1a18', marginBottom: 4, marginTop: 20 },
  authSub:         { fontSize: 13, color: '#888', marginBottom: 28 },
  fieldLabel:      { fontSize: 12, color: '#888', marginBottom: 6, fontWeight: '500' },
  input:           { backgroundColor: '#f5f5f0', borderRadius: 10, borderWidth: 0.5,
                     borderColor: '#ddd', padding: 13, fontSize: 15, color: '#1a1a18', marginBottom: 6 },
  otpInput:        { textAlign: 'center', fontSize: 24, letterSpacing: 10, fontWeight: '600' },
  hint:            { fontSize: 11, color: '#aaa', marginBottom: 20 },
  btn:             { backgroundColor: '#1D9E75', borderRadius: 12, padding: 15, alignItems: 'center', marginTop: 8 },
  btnDisabled:     { opacity: 0.6 },
  btnText:         { color: '#fff', fontSize: 15, fontWeight: '600' },
  backBtn:         { marginBottom: 16 },
  backText:        { color: '#185FA5', fontSize: 14 },
  errorText:       { backgroundColor: '#FCEBEB', color: '#791F1F', padding: 10,
                     borderRadius: 8, marginBottom: 16, fontSize: 13 },
})
