// @ghmc/mobile/src/services/api.ts
import axios, { AxiosInstance, AxiosError } from 'axios'
import * as SecureStore from 'expo-secure-store'
import AsyncStorage from '@react-native-async-storage/async-storage'

const API_URL = process.env.EXPO_PUBLIC_API_URL || 'http://localhost:3001'

// ─── Token management ─────────────────────────────────────────
export const tokenStore = {
  get:   ()            => SecureStore.getItemAsync('ghmc_token'),
  set:   (t: string)   => SecureStore.setItemAsync('ghmc_token', t),
  clear: ()            => SecureStore.deleteItemAsync('ghmc_token'),
}

// ─── Offline queue (AsyncStorage — works in Expo Go) ─────────
const QUEUE_KEY   = 'ghmc_offline_queue'
const CACHE_KEY   = 'ghmc_voter_cache'

async function queueOffline(id: string, endpoint: string, payload: any) {
  const raw  = await AsyncStorage.getItem(QUEUE_KEY)
  const queue: any[] = raw ? JSON.parse(raw) : []
  queue.push({ id, endpoint, method: 'POST', payload, created_at: Date.now(), attempts: 0 })
  await AsyncStorage.setItem(QUEUE_KEY, JSON.stringify(queue))
}

async function getQueue(): Promise<any[]> {
  const raw = await AsyncStorage.getItem(QUEUE_KEY)
  return raw ? JSON.parse(raw) : []
}

async function markSynced(id: string) {
  const queue = await getQueue()
  const updated = queue.filter(q => q.id !== id)
  await AsyncStorage.setItem(QUEUE_KEY, JSON.stringify(updated))
}

export async function getPendingCount(): Promise<number> {
  const q = await getQueue()
  return q.length
}

export async function getCachedHouseholds(boothId: string): Promise<any[]> {
  const raw = await AsyncStorage.getItem(`${CACHE_KEY}_${boothId}`)
  return raw ? JSON.parse(raw) : []
}

async function cacheHouseholds(boothId: string, data: any[]) {
  await AsyncStorage.setItem(`${CACHE_KEY}_${boothId}`, JSON.stringify(data))
}

// ─── Axios factory ────────────────────────────────────────────
async function client(): Promise<AxiosInstance> {
  const token = await tokenStore.get()
  const instance = axios.create({
    baseURL: API_URL,
    timeout: 12000,
    headers: {
      'Content-Type': 'application/json',
      ...(token ? { Authorization: `Bearer ${token}` } : {})
    }
  })
  instance.interceptors.response.use(
    res => res,
    async (err: AxiosError) => {
      if (err.response?.status === 401) await tokenStore.clear()
      return Promise.reject(err)
    }
  )
  return instance
}

// ─── API methods ──────────────────────────────────────────────
export const api = {

  // Auth
  requestOtp: async (phone: string, tenantSlug: string) => {
    const c = await client()
    const { data } = await c.post('/auth/request-otp', { phone, tenantSlug })
    return data
  },

  verifyOtp: async (phone: string, otp: string, tenantSlug: string) => {
    const c = await client()
    const { data } = await c.post('/auth/verify-otp', { phone, otp, tenantSlug })
    if (data.token) await tokenStore.set(data.token)
    return data
  },

  logout: async () => {
    await tokenStore.clear()
    await AsyncStorage.removeItem(QUEUE_KEY)
  },

  // Voters
  getVoters: async (params: Record<string, any> = {}) => {
    const c = await client()
    const { data } = await c.get('/api/voters', { params })
    return data
  },

  getVoter: async (id: string) => {
    const c = await client()
    const { data } = await c.get(`/api/voters/${id}`)
    return data
  },

  getVoterHousehold: async (id: string) => {
    const c = await client()
    const { data } = await c.get(`/api/voters/${id}/household`)
    return data
  },

  updateVoter: async (id: string, payload: any) => {
    const c = await client()
    const { data } = await c.put(`/api/voters/${id}`, payload)
    return data
  },

  // Households
  getHouseholds: async (boothId: string, visited?: boolean) => {
    const c = await client()
    const params: any = {}
    if (visited !== undefined) params.visited = visited
    const { data } = await c.get(`/api/booths/${boothId}/households`, { params })
    await cacheHouseholds(boothId, data.households || [])
    return data
  },

  // Canvassing
  logVisit: async (payload: any) => {
    try {
      const c = await client()
      const { data } = await c.post('/api/canvassing/log', payload)
      return { ...data, offline: false }
    } catch (err: any) {
      if (!err.response) {
        const localId = `offline_${Date.now()}_${Math.random().toString(36).slice(2)}`
        await queueOffline(localId, '/api/canvassing/log', payload)
        return { log_id: localId, voters_updated: 0, offline: true }
      }
      throw err
    }
  },

  syncOfflineLogs: async () => {
    const queue = await getQueue()
    const results = []
    for (const item of queue) {
      try {
        const c = await client()
        await c.post(item.endpoint, item.payload)
        await markSynced(item.id)
        results.push({ id: item.id, status: 'synced' })
      } catch {
        results.push({ id: item.id, status: 'failed' })
      }
    }
    return { results, synced: results.filter(r => r.status === 'synced').length }
  },

  getPendingCount,

  getCanvassingLogs: async (params?: any) => {
    const c = await client()
    const { data } = await c.get('/api/canvassing/logs', { params })
    return data
  },

  getMyStats: async () => {
    const c = await client()
    const { data } = await c.get('/api/canvassing/my-stats')
    return data
  },

  getSummary: async () => {
    const c = await client()
    const { data } = await c.get('/api/canvassing/summary')
    return data
  },

  // Tasks
  getTasks: async () => {
    const c = await client()
    const { data } = await c.get('/api/tasks')
    return data
  },

  updateTaskStatus: async (id: string, status: string) => {
    const c = await client()
    const { data } = await c.patch(`/api/tasks/${id}/status`, { status })
    return data
  },

  // Events
  getEvents: async () => {
    const c = await client()
    const { data } = await c.get('/api/events')
    return data
  },

  getTodayEvents: async () => {
    const c = await client()
    const { data } = await c.get('/api/events/today')
    return data
  },

  // Reports
  getCoverage: async () => {
    const c = await client()
    const { data } = await c.get('/api/reports/coverage')
    return data
  },

  getIssues: async () => {
    const c = await client()
    const { data } = await c.get('/api/reports/issues')
    return data
  },

  getVolunteerActivity: async () => {
    const c = await client()
    const { data } = await c.get('/api/reports/volunteer-activity')
    return data
  },
}
