// @ghmc/mobile/src/store/authStore.ts
import { create } from 'zustand'
import { api, tokenStore } from '../services/api'

interface User {
  id:               string
  name:             string
  role:             'tenant_owner' | 'ward_admin' | 'volunteer' | 'viewer'
  tenantSlug:       string
  tenantName:       string
  assignedWardId:   string | null
  assignedBoothId:  string | null
  schema:           string
}

interface AuthState {
  user:         User | null
  token:        string | null
  isLoading:    boolean
  error:        string | null
  tenantSlug:   string
  setTenantSlug:  (slug: string) => void
  requestOtp:     (phone: string, slug: string) => Promise<void>
  verifyOtp:      (phone: string, otp: string) => Promise<void>
  logout:         () => Promise<void>
  restoreSession: () => Promise<void>
}

// React Native safe base64 decode for JWT payload
function decodeJwtPayload(token: string): any {
  try {
    const base64 = token.split('.')[1]
      .replace(/-/g, '+')
      .replace(/_/g, '/')
    const padded = base64.padEnd(base64.length + (4 - base64.length % 4) % 4, '=')
    const decoded = Buffer.from(padded, 'base64').toString('utf8')
    return JSON.parse(decoded)
  } catch {
    return null
  }
}

export const useAuthStore = create<AuthState>((set, get) => ({
  user:       null,
  token:      null,
  isLoading:  false,
  error:      null,
  tenantSlug: '',

  setTenantSlug: (slug) => set({ tenantSlug: slug, error: null }),

  requestOtp: async (phone, slug) => {
    set({ isLoading: true, error: null })
    try {
      await api.requestOtp(phone, slug)
      set({ isLoading: false })
    } catch (err: any) {
      set({ isLoading: false, error: err.response?.data?.error || 'Failed to send OTP' })
      throw err
    }
  },

  verifyOtp: async (phone, otp) => {
    const { tenantSlug } = get()
    set({ isLoading: true, error: null })
    try {
      const data = await api.verifyOtp(phone, otp, tenantSlug)
      const u = data.user
      set({
        isLoading: false,
        token: data.token,
        user: {
          id:              u.id,
          name:            u.name,
          role:            u.role,
          tenantSlug:      u.tenantSlug,
          tenantName:      u.tenantName || u.tenantSlug,
          assignedWardId:  u.assigned_ward_id,
          assignedBoothId: u.assigned_booth_id,
          schema:          u.schema,
        }
      })
    } catch (err: any) {
      set({ isLoading: false, error: err.response?.data?.error || 'OTP verification failed' })
      throw err
    }
  },

  logout: async () => {
    await api.logout()
    set({ user: null, token: null, error: null })
  },

  restoreSession: async () => {
    const token = await tokenStore.get()
    if (!token) return
    try {
      const payload = decodeJwtPayload(token)
      if (!payload || payload.exp * 1000 < Date.now()) {
        await tokenStore.clear()
        return
      }
      set({
        token,
        user: {
          id:              payload.user_id,
          name:            payload.name,
          role:            payload.role,
          tenantSlug:      payload.tenant_slug,
          tenantName:      payload.tenant_slug,
          assignedWardId:  payload.assigned_ward_id,
          assignedBoothId: payload.assigned_booth_id,
          schema:          payload.schema,
        }
      })
    } catch {
      await tokenStore.clear()
    }
  }
}))
